p2p/background/router.go

```diff
@@ -166,6 +195,219 @@ func (rtr *backgroundRouter) RemovePeer(peer typesP2P.Peer) error {
 	return rtr.pstore.RemovePeer(peer.GetAddress())
 }
 
+// setupUnicastRouter configures and assigns `rtr.UnicastRouter`.
+func (rtr *backgroundRouter) setupUnicastRouter() error {
+	unicastRouterCfg := config.UnicastRouterConfig{
+		Logger:         rtr.logger,
+		Host:           rtr.host,
+		ProtocolID:     protocol.BackgroundProtocolID,
+		MessageHandler: rtr.handleBackgroundMsg,
+		PeerHandler:    rtr.AddPeer,
+	}
+
+	unicastRouter, err := unicast.Create(rtr.GetBus(), &unicastRouterCfg)
+	if err != nil {
+		return fmt.Errorf("setting up unicast router: %w", err)
+	}
+
+	rtr.UnicastRouter = *unicastRouter
+	return nil
+}
+
+func (rtr *backgroundRouter) setupDependencies(ctx context.Context, cfg *config.BackgroundConfig) error {
+	if err := rtr.setupUnicastRouter(); err != nil {
+		return err
+	}
+
+	if err := rtr.setupPeerDiscovery(ctx); err != nil {
+		return fmt.Errorf("setting up peer discovery: %w", err)
+	}
+
+	if err := rtr.setupPubsub(ctx); err != nil {
+		return fmt.Errorf("setting up pubsub: %w", err)
+	}
+
+	if err := rtr.setupTopic(); err != nil {
+		return fmt.Errorf("setting up topic: %w", err)
+	}
+
+	if err := rtr.setupSubscription(); err != nil {
+		return fmt.Errorf("setting up subscription: %w", err)
+	}
+
+	if err := rtr.setupPeerstore(
+		ctx,
+		cfg.PeerstoreProvider, cfg.CurrentHeightProvider); err != nil {
+		return fmt.Errorf("setting up peerstore: %w", err)
+	}
+	return nil
+}
+
+func (rtr *backgroundRouter) setupPeerstore(
+	ctx context.Context,
+	pstoreProvider providers.PeerstoreProvider,
+	currentHeightProvider providers.CurrentHeightProvider,
+) (err error) {
+	// seed initial peerstore with current on-chain peer info (i.e. staked actors)
+	rtr.pstore, err = pstoreProvider.GetStakedPeerstoreAtHeight(
+		currentHeightProvider.CurrentHeight(),
+	)
+	if err != nil {
+		return err
+	}
+
+	if err := rtr.bootstrap(ctx); err != nil {
+		return fmt.Errorf("bootstrapping peerstore: %w", err)
+	}
+
+	return nil
+}
+
+func (rtr *backgroundRouter) setupPeerDiscovery(ctx context.Context) (err error) {
+	dhtMode := dht.ModeAutoServer
+	// NB: don't act as a bootstrap node in peer discovery in client debug mode
+	if isClientDebugMode(rtr.GetBus()) {
+		dhtMode = dht.ModeClient
+	}
+
+	rtr.kadDHT, err = dht.New(ctx, rtr.host, dht.Mode(dhtMode))
+	return err
+}
+
+func (rtr *backgroundRouter) setupPubsub(ctx context.Context) (err error) {
+	// TECHDEBT(#730): integrate libp2p tracing via `pubsub.WithEventTracer()`.
+
+	// CONSIDERATION: If switching to `NewRandomSub`, there will be a max size
+	rtr.gossipSub, err = pubsub.NewGossipSub(ctx, rtr.host)
+	return err
+}
+
+func (rtr *backgroundRouter) setupTopic() (err error) {
```

Comment:
Why are we using named `err` return values in this file?

Reply:
To simplify line 296:

```go
if rtr.topic, err = ...
```

It saves a line. I think there are valid scenarios where we definitely do want named return args; e.g. when they're ambiguous or if we need to trace an error in a defer. While this isn't one of those scenarios, that's what makes me feel comfortable converting this to one.

(I **do not** condone "naked returns" though.)